### PR TITLE
[Fix] for Issue #557 moving focus to the keyboard shortcuts panel when opened

### PR DESCRIPTION
**Pull request by erin-doyle**
Modified aria attributes and focusing behavior so that the keyboard shortcuts panel receives focus, is read appropriately by screen readers, prevents focus from accidentally moving back to the other elements without use of the proper shortcuts, and then returns focus to the appropriate place when closed.

I tested this on Chrome and Safari with Mac VoiceOver and FireFox with no screen reader.  It seems that all the keyboard navigation and screen reading is working consistently across them all now.  Triggering the keyboard shortcuts panel opens, focuses and returns focus when closed appropriately now regardless of how its triggered.

_EDIT:_ also tested in Microsoft Edge along with NVDA.

### CODE REVIEW

File: src/components/DayPicker.jsx [hunk 1]
```diff
@@ -234,13 +234,10 @@ class DayPicker extends React.Component {
   }
 
   componentDidUpdate(prevProps) {
-    const { isFocused, showKeyboardShortcuts } = this.props;
+    const { isFocused } = this.props;
     const { focusedDate } = this.state;
 
-    if (
-      (!prevProps.isFocused && isFocused && !focusedDate) ||
-      (!prevProps.showKeyboardShortcuts && showKeyboardShortcuts)
```
Comment (erin-doyle) [inline src/components/DayPicker.jsx]: I removed this line because it was pulling the focus back to the DayPicker after it had just been moved to the keyboard shortcuts panel.

File: src/components/KeyboardShortcutRow.jsx [hunk 2]
```diff
@@ -38,7 +38,7 @@ function KeyboardShortcutRow({
         <span
           {...css(styles.KeyboardShortcutRow_key)}
           role="img"
-          aria-label={label}
+          aria-label={`${label} `}
```
Comment (erin-doyle) [inline src/components/KeyboardShortcutRow.jsx]: Adding in the space here inserts a space between the screen reader reading the label and then the {action} from the <div> below.  Previously it was reading like:

"Enter keySelect the date in focus"

Basically it was merging the last word of one to the first word of the next and making it hard to understand when read.

File: src/components/SingleDatePicker.jsx [hunk 3]
```diff
@@ -404,6 +404,7 @@ class SingleDatePicker extends React.Component {
           renderCalendarInfo={renderCalendarInfo}
           isFocused={isDayPickerFocused}
           showKeyboardShortcuts={showKeyboardShortcuts}
+          onBlur={this.onDayPickerBlur}
```
Comment (erin-doyle) [inline src/components/SingleDatePicker.jsx]: This was another important miss I found in functionality between the DateRangePicker but not the SingleDatePicker.  This was the missing piece to the focus returning back to the DateInput when that was where the keyboard shortcuts panel was triggered to open.

File: src/components/DayPickerKeyboardShortcuts.jsx [hunk 4]
```diff
@@ -33,15 +33,91 @@ const defaultProps = {
   phrases: DayPickerKeyboardShortcutsPhrases,
 };
 
+function getKeyboardShortcuts(phrases) {
+  return [{
+    unicode: '↵',
+    label: phrases.enterKey,
+    action: phrases.selectFocusedDate,
+  },
```
Comment (ljharb) [inline src/components/DayPickerKeyboardShortcuts.jsx]: the indentation here is a little off; if it's going to start with `[{`, then each line needs to be `}, {` - otherwise the first `{` needs to be on its own line.

File: src/components/DayPickerKeyboardShortcuts.jsx [hunk 5]
```diff
@@ -33,15 +33,91 @@ const defaultProps = {
...
+  ];
+}
+
 class DayPickerKeyboardShortcuts extends React.Component {
   constructor(...args) {
     super(...args);
 
-    this.onClick = this.onClick.bind(this);
+    this.keyboardShortcuts = getKeyboardShortcuts(this.props.phrases);
```
Comment (ljharb) [inline src/components/DayPickerKeyboardShortcuts.jsx]: this would also need to happen in `componentWillReceiveProps`.

Comment (erin-doyle) [reply src/components/DayPickerKeyboardShortcuts.jsx]: Moving this was not performance motivated, I just wanted to make render() a little more concise.  Sounds like I should just move it back.

Comment (ljharb) [reply src/components/DayPickerKeyboardShortcuts.jsx]: I think it's a great idea to move it; but if it's not in the render path, then it has to be updated in both of the places where props change.

Comment (erin-doyle) [reply src/components/DayPickerKeyboardShortcuts.jsx]: Ok I'll make that change in a separate changeset then 👍

File: src/components/DayPickerKeyboardShortcuts.jsx [hunk 6]
```diff
@@ -120,7 +168,14 @@ class DayPickerKeyboardShortcuts extends React.Component {
           )}
           type="button"
           aria-label={toggleButtonText}
-          onClick={this.onClick}
+          onClick={this.onShowKeyboardShortcutsButtonClick}
+          onKeyDown={(e) => {
+            if (e.key === 'Enter') {
+              e.preventDefault();
+            } else if (e.key === 'Space') {
+              this.onShowKeyboardShortcutsButtonClick();
```
Comment (ljharb) [inline src/components/DayPickerKeyboardShortcuts.jsx]: this should probably pass the `e` through?

File: src/components/DayPickerKeyboardShortcuts.jsx [hunk 7]
```diff
@@ -110,6 +157,7 @@ class DayPickerKeyboardShortcuts extends React.Component {
     return (
       <div>
         <button
+          id="showKeyboardShortcutsButton"
```
Comment (ljharb) [inline src/components/DayPickerKeyboardShortcuts.jsx]: it seems like this ID only exists so it can be targeted in tests; if so, please remove it - there's surely another way it can be located.

File: src/components/DayPickerKeyboardShortcuts.jsx [hunk 8]
```diff
@@ -37,11 +37,46 @@ class DayPickerKeyboardShortcuts extends React.Component {
...
 
-  onClick() {
+  componentDidUpdate() {
+    this.handleFocus();
+  }
+
+  onKeyDown(e) {
+    const { closeKeyboardShortcutsPanel } = this.props;
+
+    e.stopPropagation();
```
Comment (ljharb) [inline src/components/DayPickerKeyboardShortcuts.jsx]: why does propagation need to be stopped here?

File: src/components/DayPickerKeyboardShortcuts.jsx [hunk 9]
```diff
@@ -154,26 +210,25 @@ class DayPickerKeyboardShortcuts extends React.Component {
             </div>
 
             <button
+              id="hideKeyboardShortcutsButton"
```
Comment (ljharb) [inline src/components/DayPickerKeyboardShortcuts.jsx]: also this one

File: src/components/DayPickerKeyboardShortcuts.jsx [hunk 10]
```diff
@@ -154,26 +210,25 @@ class DayPickerKeyboardShortcuts extends React.Component {
...
-              }}
+              onKeyDown={this.onKeyDown}
             >
               <CloseButton {...css(styles.DayPickerKeyboardShortcuts_closeSvg)} />
             </button>
 
-            <ul {...css(styles.DayPickerKeyboardShortcuts__list)}>
+            <ul
+              {...css(styles.DayPickerKeyboardShortcuts__list)}
+              id="DayPickerKeyboardShortcuts__description"
```
Comment (ljharb) [inline src/components/DayPickerKeyboardShortcuts.jsx]: i'm not sure if this is a pattern we already follow or not; but an ID must be unique on the page, and there should be able to be multiple date of these in the DOM at once - is there another way this can be linked?

File: src/components/KeyboardShortcutRow.jsx (hunk 2 above)
Comment (ljharb) [reply src/components/KeyboardShortcutRow.jsx]: would adding `:` or similar instead of a space have the same effect?

File: src/components/DayPickerKeyboardShortcuts.jsx (hunk 8 above)
Comment (erin-doyle) [reply src/components/DayPickerKeyboardShortcuts.jsx]: It's being done that way in DayPicker's `onKeyDown()` so figured it would probably be needed here as well.

File: src/components/DayPickerKeyboardShortcuts.jsx (hunk 7 above)
Comment (erin-doyle) [reply src/components/DayPickerKeyboardShortcuts.jsx]: Ok I'll remove and update the test(s)

File: src/components/DayPickerKeyboardShortcuts.jsx (hunk 6 above)
Comment (ljharb) [reply src/components/DayPickerKeyboardShortcuts.jsx]: i think it'd be more future-proof if it ever needed it in the future. Not a huge deal obv.

File: src/components/KeyboardShortcutRow.jsx (hunk 2 above)
Comment (erin-doyle) [reply src/components/KeyboardShortcutRow.jsx]: Adding a `:` works great with Safari but in Chrome the punctuation is read so it sounds like:

"Enter key colon Select the date in focus"

I tried a `,` instead and that seems to work nicely and consistently in both browsers.  So I'll change this to a `,` instead of a space

Comment (ljharb) [reply src/components/KeyboardShortcutRow.jsx]: awesome!

**Comment by erin-doyle** [comment]
I just want to follow up on this and make sure there weren't any more outstanding actions on my part that you may be waiting for.  I think I got it all but if not let me know!

File: src/components/DayPicker.jsx (hunk 1 above)
Comment (majapw) [reply src/components/DayPicker.jsx]: hmmm, i don't remember. Focusing in the keyboard shortcuts panel may have evolved since I first integrated it and this may be a relic of the initial a11y change.

File: src/components/DayPickerKeyboardShortcuts.jsx [hunk 11]
```diff
@@ -97,7 +70,98 @@ class DayPickerKeyboardShortcuts extends React.Component {
...
+
+      case 'Tab':
+      case 'Enter':
+      case 'Home':
+      case 'End':
+      case 'PageUp':
+      case 'PageDown':
+      case 'ArrowLeft':
+      case 'ArrowRight':
+        e.stopPropagation();
```
Comment (majapw) [inline src/components/DayPickerKeyboardShortcuts.jsx]: Can you clarify why you're stopping propagation here and below? and preventing default?

File: src/components/DayPickerKeyboardShortcuts.jsx [hunk 12]
```diff
@@ -97,7 +70,98 @@ class DayPickerKeyboardShortcuts extends React.Component {
...
+        e.preventDefault();
+        break;
+
+      case 'ArrowUp':
+      case 'ArrowDown':
+        e.stopPropagation();
```
Comment (majapw) [inline src/components/DayPickerKeyboardShortcuts.jsx]: Can you leave a comment in here to explain what's going on? I'm not sure it's self-evident

File: src/components/KeyboardShortcutRow.jsx [hunk 13]
```diff
@@ -38,7 +38,7 @@ function KeyboardShortcutRow({
         <span
           {...css(styles.KeyboardShortcutRow_key)}
           role="img"
-          aria-label={label}
+          aria-label={`${label},`}
```
Comment (majapw) [inline src/components/KeyboardShortcutRow.jsx]: ?

Comment (ljharb) [reply src/components/KeyboardShortcutRow.jsx]: https://github.com/airbnb/react-dates/pull/825#discussion_r148938046

Comment (erin-doyle) [reply src/components/KeyboardShortcutRow.jsx]: I can add a comment to the code on this as well.

File: src/components/DayPickerKeyboardShortcuts.jsx (hunk 12 above)
Comment (erin-doyle) [reply src/components/DayPickerKeyboardShortcuts.jsx]: Yep no problem!

**Comment by erin-doyle** [comment]
Looking in to build failures now, all I did was rebase on master to pick up any recent changes and then add comments...

**Comment by erin-doyle** [comment]
Linting error:

111:1  error  Line 111 exceeds the maximum line length of 100  max-len

:woman_facepalming:

**Comment by erin-doyle** [comment]
Yay!  🎉 😌

**Comment by erin-doyle** [comment]
Would it be cool to go ahead and push a version update now that this is in?  Patch?  I can do it if y'all are busy but didn't want to assume in case you were holding off for something on its way.

**Comment by majapw** [comment]
v15.1.0 is out!
